lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -591,30 +597,41 @@ public void OnNext(IFailedEvaluator failedEvaluator)
         /// <param name="failedContext"></param>
         public void OnNext(IFailedContext failedContext)
         {
+            Logger.Log(Level.Warning, "Received IFailedContext with Id: {0} from endpoint {1} with systemState {2} in retry#: {3}.", failedContext.Id, GetEndPointFromContext(failedContext), _systemState.CurrentState, _numberOfRetries);
             lock (_lock)
             {
-                if (_taskManager.AreAllTasksCompleted())
+                using (Logger.LogFunction("IMRUDriver::IFailedContext"))
                 {
-                    Logger.Log(Level.Info, "Context with Id: {0} failed but IMRU tasks are completed. So ignoring.", failedContext.Id);
-                    return;
-                }
+                    switch (_systemState.CurrentState)
```

Comment:
What about `WaitingForEvaluator` state - theoretically we could get a failure in it as well... I'd make throwing exception `default` clause, so that we only list `TasksCompleted`, `ShuttingDown` and `Fail` explicitly